### PR TITLE
Add hue traits

### LINKED ISSUE #12200: Add hue rotation traits
Nit for future PR: We should add hue rotation traits to make this cleaner.

_Originally posted by @bushrat011899 in https://github.com/bevyengine/bevy/pull/12163#discussion_r1506997793_
            
            
Much like the `Alpha` trait, we should provide easy ways to "rotate" the hue of colors: shifting them by a provided number of degrees to rotate the metaphorical color wheel.

This trait should only be implemented for the color models (types) with an explicit hue element: other color types should be converted to one of those spaces first.

Once basic hue rotation is added, we can add nice color theory options: things like "complementary colors" (rotate by 180 degrees), "triadic colors" (rotate by 120 degrees), analogous colors and so on.

### PR DESCRIPTION
**Pull request by oyasumi731**
# Objective

Fixes #12200 .

## Solution

I added a Hue Trait with the rotate_hue method to enable hue rotation.
Additionally, I modified the implementation of animations in the animated_material sample.

---

## Changelog

- Added a  `Hue` trait to `bevy_color/src/color_ops.rs`.
- Added  the `Hue` trait implementation to `Hsla`, `Hsva`, `Hwba`, `Lcha`, and `Oklcha`.
- Updated animated_material sample.

## Migration Guide

Users of Oklcha need to change their usage to use the with_hue method instead of the with_h method.

### CODE REVIEW

File: crates/bevy_color/src/color_ops.rs [hunk 1]
```diff
@@ -61,3 +61,39 @@ pub trait Alpha: Sized {
         self.alpha() >= 1.0
     }
 }
+
+/// Method for manipulating the hue of a color.
```
Comment (pablo-lua) [inline crates/bevy_color/src/color_ops.rs:65]:
```suggestion
/// Trait for manipulating the hue of a color.
```

File: crates/bevy_color/src/color_ops.rs [hunk 2]
```diff
@@ -61,3 +61,39 @@ pub trait Alpha: Sized {
...
+pub trait Hue: Sized {
+    /// Return a new version of this color with the hue channel set to the given value.
+    fn with_hue(&self, hue: f32) -> Self;
+
+    /// Return the hue of this color (0.0 - 360.0).
```
Comment (pablo-lua) [inline crates/bevy_color/src/color_ops.rs:70]:
```suggestion
    /// Return the hue of this color [0.0, 360.0].
```

Comment (pablo-lua) [reply crates/bevy_color/src/color_ops.rs:70]: At the same time, I think we shouldn't keep the interval here. What if we find a colorspace that for some reason will go by another interval?

**Review by pablo-lua — APPROVED** [review]
I have just some comments here. But everything else looks fine, thanks for tackling this!

File: crates/bevy_color/src/color_ops.rs (hunk 2 above)
Comment (oyasumi731) [reply crates/bevy_color/src/color_ops.rs]: Certainly, it's not guaranteed that the range of hue values will fit within this range in every color space.
Thank you for your comment!

Comment (pablo-lua) [reply crates/bevy_color/src/color_ops.rs]: Yeah, I agree that we don't have many colorspaces out there with some odd behavior, but  at the same time, this trait is made to be generic, and I think that should reflect on docs

File: examples/3d/animated_material.rs [hunk 3]
```diff
@@ -30,14 +30,16 @@ fn setup(
...
     for x in -1..2 {
         for z in -1..2 {
             commands.spawn(PbrBundle {
                 mesh: cube.clone(),
-                material: materials.add(Color::WHITE),
+                material: materials.add(Color::hsl((i as f32 * 2.345) * 100.0 % 360.0, 1.0, 0.5)),
                 transform: Transform::from_translation(Vec3::new(x as f32, 0.0, z as f32)),
                 ..default()
             });
+            i += 1;
```
Comment (MiniaczQ) [inline examples/3d/animated_material.rs:42]: Not important, but you can use the `rotate_hue` and a cool property of rotating by golden ratio here for showcase:
```rs
// Not a stable f32 constant yet
const PHI: f32 = 1.6180339;

let mut c = Color::hsl(0.0, 1.0, 0.5);
for x in -1..2 {
  for z in -1..2 {
    // Pbr spawning
    c = c.rotate_hue(PHI);
  }
}
```

Comment (oyasumi731) [reply examples/3d/animated_material.rs]: As you suggested, rotating at a constant ratio makes for a smarter implementation indeed. I rotated it at the golden angle to ensure that the colors of adjacent materials diverge!

File: crates/bevy_color/src/color_ops.rs [hunk 4]
```diff
@@ -61,3 +61,39 @@ pub trait Alpha: Sized {
...
+    /// Return the hue of this color [0.0, 360.0].
+    fn hue(&self) -> f32;
+
+    /// Sets the hue of this color.
+    fn set_hue(&mut self, hue: f32);
+
+    /// Return a new version of this color with the hue channel rotated by the given degrees.
+    fn rotate_hue(&self, degrees: f32) -> Self {
+        let degrees = degrees.rem_euclid(360.);
+        self.with_hue((self.hue() + degrees) % 360.)
```
Comment (viridia) [inline crates/bevy_color/src/color_ops.rs:79]: Why not call `.rem_euclid()` on the result after the addition?

**Review by viridia — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
@oyasumi731 I've done my best to resolve merge conflicts: going to try and merge this :) If it fails, please clean up my changes and get CI passing and then ping me to retry the merge.

**Comment by oyasumi731** [comment]
@alice-i-cecile Thank you.
I fixed the CI errors, so please merge again.

**Comment by alice-i-cecile** [comment]
Thank you very much :) I really appreciate when authors are able to fix up things like that for us.

**Pull request merged**